listenbrainz/webserver/static/js/src/types.d.ts

```diff
@@ -46,6 +46,17 @@ declare type Listen = {
   score?: number;
 };
 
+declare type Recommendation = {
```

Comment:
Unless I'm missing something, this should be equivalent to the Listens type declared above, since the fields you removed (listened_at_iso, playing_now) are marked as optional (the '?' after the variable name).
You should be able to use the Listen type directly without breaking anything.

Reply:
@MonkeyDo I dont think rec will have listened_at. So we went for separate component. @vansika can you remove that

Reply:
@MonkeyDo I am trying to keep listens and recommendations as much separate as I can. But BP accepts `Listens` type. Like @shivam-kapila said, recs won't have a listened_at field either but `Listens` type makes it mandatory to have listened at, here https://github.com/metabrainz/listenbrainz-server/blob/master/listenbrainz/webserver/static/js/src/types.d.ts#L36

So when I remove `listened_at` from recs I get errors saying `Listened_at missing`, that is why I have kept listenes_at in recs type until I figure out how to solve it. 

Ideally the recs type should be:
`declare type Recommendation = {
  user_name?: string | null;
  track_metadata: {
    artist_name: string;
    release_name?: string | null;
    track_name: string;
    additional_info?: AdditionalInfo;
  };
};`

Reply:
To me that sounds more like an adjustment needed on the Listens type, until there is a clear need to make another type, to avoid code duplication (especially considering the recommendations are based on the listens format).
Namely to make `listened_at` optional: `listened_at?: number` (note the question mark which makes it equal to `number | undefined`).

Inevitably there will be a couple of changes to make to other components, where typescript will complain that `listened_at is possibly 'undefined'.`.
These we should be able to solve relatively easily, for example transforming `new Date(listen.listened_at * 1000).toISOString()` into `(listen.listened_at ? new Date(listen.listened_at * 1000).toISOString() : null)` to ensure listened_at is defined.




There is also another way: [extending](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html) the Listen type, or creating a BaseListen type and extending it for Listen and Recommendation types:
```typescript
declare type Listen = {
  listened_at?: number;
  listened_at_iso?: string | null;
  playing_now?: boolean | null;
  user_name?: string | null;
  track_metadata: {
    artist_name: string;
    release_name?: string | null;
    track_name: string;
    additional_info?: AdditionalInfo;
  };
};

declare type Recommendation = Listen & {
  score?: number;
};
```

Or:
```typescript
declare type BaseListenFormat = {
  user_name?: string | null;
  track_metadata: {
    artist_name: string;
    release_name?: string | null;
    track_name: string;
    additional_info?: AdditionalInfo;
  };
};

declare type Listen = BaseListenFormat & {
  listened_at?: number;
  listened_at_iso?: string | null;
  playing_now?: boolean | null;
};

declare type Recommendation = BaseListenFormat & {
  score?: number;
};
```

Reply:
I like the idea of making listened_at optional (there will be tests for `my_listens` etc to ensure that listened_at *must* be passed with Listens so it will work :)). 

is there a way to rename listens -> recommendations when they will be used in Recommendations.tsx? I am trying to avoid Listens name in Recommendation component.

Reply:
`declare type Listen = {
  listened_at?: number;
  listened_at_iso?: string | null;
  playing_now?: boolean | null;
  user_name?: string | null;
  track_metadata: {
    artist_name: string;
    release_name?: string | null;
    track_name: string;
    additional_info?: AdditionalInfo;
  };
};

declare type Recommendation = Listen

Reply:
How about this? (Score isn't required)

Reply:
That could work.
I think in that case you could just rename the Listen type instead to something that will work with both cases :)

But I think my second suggestion (extend BaseListenFormat into Listen and Recommendation types) would solve that issue and allow for things to change a bit more easily in the future.

Reply:
I dont know what you are trying to achieve by changing RecentListens but I used and all worked fine 

```
  user_name?: string | null;
  track_metadata: {
    artist_name: string;
    release_name?: string | null;
    track_name: string;
    additional_info?: AdditionalInfo;
  };
};

declare type Listen = BaseListenFormat & {
  listened_at: number;
  listened_at_iso?: string | null;
  playing_now?: boolean | null;
};

declare type Recommendation = BaseListenFormat;
```